pkg/daemonset/daemonset_test.go

```diff
@@ -52,3 +65,391 @@ func TestWithOptions(t *testing.T) {
+	testCases := []struct {
+		name          string
+		namespace     string
+		nodeLabels    map[string]string
+		containerSpec corev1.Container
+		expectedError string
+		apiClientNil  bool
+	}{
+		{ // Test case 1 - API client is nil
+			name:      "test-name",
+			namespace: "test-namespace",
+			nodeLabels: map[string]string{
+				"test-key": "test-value",
+			},
+			containerSpec: corev1.Container{
+				Name: "test-container",
+			},
+			expectedError: "",
+			apiClientNil:  true,
+		},
+		{ // Test case 2 - name is empty
+			name:          "",
+			namespace:     "test-namespace",
+			nodeLabels:    map[string]string{},
+			containerSpec: corev1.Container{},
+			expectedError: "daemonset 'name' cannot be empty",
+			apiClientNil:  false,
+		},
+		{ // Test case 3 - namespace is empty
+			name:          "test-name",
+			namespace:     "",
+			nodeLabels:    map[string]string{},
+			containerSpec: corev1.Container{},
+			expectedError: "daemonset 'namespace' cannot be empty",
+			apiClientNil:  false,
+		},
+		{ // Test case 4 - API client is not nil
+			name:      "test-name",
+			namespace: "test-namespace",
+			nodeLabels: map[string]string{
+				"test-key": "test-value",
+			},
+			containerSpec: corev1.Container{
+				Name: "test-container",
+			},
+			expectedError: "",
+			apiClientNil:  false,
+		},
+		{ // Test case 5 - labels are empty
+			name:       "test-name",
+			namespace:  "test-namespace",
+			nodeLabels: map[string]string{},
+			containerSpec: corev1.Container{
+				Name: "test-container",
+			},
+			expectedError: "daemonset 'labels' cannot be empty",
+		},
+	}
+
+	for _, testCase := range testCases {
+		var testClients *clients.Settings
+
+		if testCase.apiClientNil {
+			testClients = nil
+		} else {
+			testClients = clients.GetTestClients(clients.TestClientParams{})
+		}
+
+		testBuilder := NewBuilder(testClients, testCase.name, testCase.namespace, testCase.nodeLabels, testCase.containerSpec)
+
+		if testCase.apiClientNil {
+			assert.Nil(t, testBuilder)
+		} else {
+			if testCase.expectedError != "" {
+				assert.Equal(t, testCase.expectedError, testBuilder.errorMsg)
+			} else {
+				assert.NotNil(t, testBuilder)
+				assert.Equal(t, testCase.name, testBuilder.Definition.Name)
+				assert.Equal(t, testCase.namespace, testBuilder.Definition.Namespace)
+				assert.Equal(t, testCase.nodeLabels, testBuilder.Definition.Spec.Selector.MatchLabels)
+				assert.Equal(t, testCase.containerSpec.Name, testBuilder.Definition.Spec.Template.Spec.Containers[0].Name)
+			}
+		}
+	}
+}
+
+func TestDaemonsetPull(t *testing.T) {
+	generateDaemonset := func(name, namespace string) *appsv1.DaemonSet {
+		return &appsv1.DaemonSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+		}
+	}
+
+	testCases := []struct {
+		name                string
+		namespace           string
+		expectedError       bool
+		addToRuntimeObjects bool
+		expectedErrorText   string
```

Comment:
Add client and verify that Pull returns error if apiClient is nil.

Reply:
Done.